packages/twenty-server/src/modules/messaging/message-folder-manager/services/sync-message-folders.service.ts

```diff
@@ -0,0 +1,189 @@
+    );
+
+    if (!messageChannel) {
+      throw new Error(`Message channel ${messageChannelId} not found`);
+    }
+
+    const folders = await this.discoverAllFolders(messageChannel);
+
+    await this.upsertDiscoveredFolders({
+      workspaceId,
+      messageChannelId,
+      folders,
+      manager,
+    });
+  }
+
+  private async upsertDiscoveredFolders({
+    workspaceId,
+    messageChannelId,
+    folders,
+    manager,
+  }: {
+    workspaceId: string;
+    messageChannelId: string;
+    folders: MessageFolder[];
+    manager: WorkspaceEntityManager;
+  }): Promise<void> {
+    const messageFolderRepository =
+      await this.twentyORMGlobalManager.getRepositoryForWorkspace<MessageFolderWorkspaceEntity>(
+        workspaceId,
+        'messageFolder',
+      );
+
+    for (const folder of folders) {
+      const existingFolder = await this.findExistingFolder({
+        messageChannelId,
+        folder,
+        messageFolderRepository,
+      });
+
+      if (existingFolder) {
+        await messageFolderRepository.update(
+          existingFolder.id,
+          {
+            name: folder.name,
+            isSynced: folder.isSynced,
+            isSentFolder: folder.isSentFolder,
+            externalId: folder.externalId,
+          },
+          manager,
+        );
+      } else {
+        await messageFolderRepository.save(
+          {
+            id: v4(),
+            messageChannelId,
+            name: folder.name,
+            syncCursor: '',
+            isSynced: folder.isSynced,
+            isSentFolder: folder.isSentFolder,
+            externalId: folder.externalId,
+          },
+          {},
+          manager,
+        );
+      }
+    }
+  }
+
+  async discoverAllFolders(
+    messageChannel: MessageChannelWorkspaceEntity,
+  ): Promise<MessageFolder[]> {
+    switch (messageChannel.connectedAccount.provider) {
+      case ConnectedAccountProvider.GOOGLE:
+        return await this.gmailGetAllFoldersService.getAllMessageFolders(
+          messageChannel.connectedAccount,
+        );
+      case ConnectedAccountProvider.MICROSOFT:
+        return await this.microsoftGetAllFoldersService.getAllMessageFolders(
+          messageChannel.connectedAccount,
+        );
+      case ConnectedAccountProvider.IMAP_SMTP_CALDAV:
+        return await this.imapGetAllFoldersService.getAllMessageFolders(
+          messageChannel.connectedAccount,
+        );
+      default:
+        throw new Error(
+          `Provider ${messageChannel.connectedAccount.provider} is not supported`,
+        );
+    }
+  }
+
+  private async findExistingFolder({
+    messageChannelId,
+    folder,
+    messageFolderRepository,
+  }: {
+    messageChannelId: string;
+    folder: MessageFolder;
+    messageFolderRepository: WorkspaceRepository<MessageFolderWorkspaceEntity>;
+  }): Promise<MessageFolderWorkspaceEntity | null> {
+    if (isDefined(folder.externalId)) {
+      const existingFolder = await messageFolderRepository.findOne({
```

Comment:
this is not efficient, this will load the database too much: we are doing a call for each folder

Reply:
let's query them all before